tree.go

```diff
@@ -552,24 +570,105 @@ func (n *InternalNode) Commitment() *Point {
 }
 
 func (n *InternalNode) Commit() *Point {
-	poly := make([]Fr, NodeWidth)
-	emptyChildren := 256
-
+	if n.depth <= 1 {
+		return n.commitRoot()
+	}
 	if len(n.cow) != 0 {
+		polyp := frPool.Get().(*[]Fr)
+		poly := *polyp
+		defer func() {
+			for i := 0; i < NodeWidth; i++ {
+				poly[i] = Fr{}
+			}
+			frPool.Put(polyp)
+		}()
+		emptyChildren := 256
+
+		var i int
+		b := make([]byte, len(n.cow))
+		points := make([]*Point, 2*len(n.cow))
 		for idx, comm := range n.cow {
 			emptyChildren--
-			var pre Fr
-			// TODO use kev's multimaptofield
-			toFr(&pre, comm)
-			// child in cow, so its child has also been
-			// modified, so call `Commit()` instead of
-			// `Commitment()`
-			toFr(&poly[idx], n.children[idx].Commit())
-			poly[idx].Sub(&poly[idx], &pre)
+			points[2*i] = comm
+			points[2*i+1] = n.children[idx].Commit()
+			b[i] = idx
+			i++
 		}
+		frs := make([]*Fr, len(points))
+		for i := range frs {
+			frs[i] = &Fr{}
+		}
+		toFrMultiple(frs, points)
+		for i := 0; i < len(points)/2; i++ {
+			poly[b[i]].Sub(frs[2*i+1], frs[2*i])
+		}
+
 		n.cow = nil
 
-		n.commitment.Add(n.commitment, GetConfig().CommitToPoly(poly, emptyChildren))
+		n.commitment.Add(n.commitment, cfg.CommitToPoly(poly, emptyChildren))
+		return n.commitment
+	}
+
+	return n.commitment
+}
+
+func (n *InternalNode) commitRoot() *Point {
+	if len(n.cow) != 0 {
```

Comment:
This is the parallel version of L576-612. The code is quite similar, but it separates the work in batches and delegates work to different goroutines.

This idea is similar to the mentioned parallel `Flush()` changed explained before.